File: 卡码/代码随想录/1/11-图论/Java/src/com/kurumi/dfs/LeetCode_0797.java
//797. 所有可能的路径
package com.kurumi.dfs;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class LeetCode_0797 {

}
class Solution {
	List<List<Integer>> res=new ArrayList<>();
	Deque<Integer> path=new ArrayDeque();
	public void dfs(int[][]graph,int index) {
		if(index==graph.length-1) {
			res.add(new ArrayList<>(path));
			return;
		}
		for(int i=0;i<graph[index].length;i++) {
			path.offer(graph[index][i]);
			dfs(graph,graph[index][i]);
			path.pollLast();
		}
	}
  public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
  	path.add(0);
  	dfs(graph,0);
  	return res;
  }
}

File: 卡码/代码随想录/1/11-图论/Go/dfs/LeetCode_0797.go
// 797. 所有可能的路径
package main

func allPathsSourceTarget(graph [][]int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)
	var dfs func(index int)
	dfs = func(index int) {
		if index == len(graph)-1 {
			tmp := make([]int, len(path))
			copy(tmp, path)
			res = append(res, tmp)
			return
		}
		for _, v := range graph[index] {
			path = append(path, v)
			dfs(v)
			path = path[:len(path)-1]
		}
	}
	path = append(path, 0)
	dfs(0)
	return res
}
